src/Database/Redis/Schema/RemoteJob.hs

```diff
@@ -0,0 +1,356 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Database.Redis.Schema.RemoteJob (
+  -- * Types
+  WorkerId (..),
+  WorkerHandle,
+  RemoteJobError (..),
+  JobQueue (..),
+
+  -- * Main functionality
+  runRemoteJob,
+  runRemoteJobAsync,
+  remoteJobWorker,
+  withRemoteJobWorker,
+  gracefulShutdown,
+
+  -- * Inspection
+  queueLength,
+  countRunningJobs,
+) where
+
+import Data.Binary ( decode, encode, Binary(..) )
+import Data.Bifunctor as BF ( second )
+import Data.Kind ( Type )
+import Data.Maybe ( isJust )
+import Data.Proxy ( Proxy(..) )
+import Data.String ( IsString )
+import Data.Text ( Text )
+import Data.Time ( UTCTime, getCurrentTime, addUTCTime )
+import Data.Time.Clock.POSIX ( utcTimeToPOSIXSeconds )
+import Data.UUID ( UUID )
+import Data.UUID.V4 ( nextRandom )
+import qualified Data.ByteString.Char8 as BS
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Set as Set
+
+import Database.Redis ( ConnectionLostException )
+import Database.Redis.Schema as Redis
+
+import Control.Concurrent.MonadIO
+import Control.Monad ( when, forever )
+import Control.Monad.Catch
+import Control.Exception ( SomeAsyncException )
+
+import GHC.Generics ( Generic )
+
+
+-- | Errors that can occur in the remote job running.
+data RemoteJobError
+  = RemoteJobException String
+  | NoActiveWorkers
+  | Timeout
+  deriving ( Show, Generic )
+instance Binary RemoteJobError
+
+instance Serializable RemoteJobError where
+  fromBS = readBinary
+  toBS = showBinary
+
+-- | Identifier for the worker process, is only used for inspecting the queue
+newtype WorkerId = WorkerId { unWorkerId :: Text }
+  deriving newtype ( Show, Eq, Ord, IsString, Serializable, Binary )
+
+-- | Handle of the worker process, which can be used in the graceful shutdown procedure
+data WorkerHandle = WorkerHandle (MVar ()) ThreadId
+
+class JobQueue jq where
+  -- | The remote job protocol, a list of 'i -> o' entries indicating
+  --   this queue can contains jobs taking type 'i' as input and
+  --   returning type 'o'. Both 'i' and 'o' must have a binary instance.
+  type RPC jq :: [Type]
+
+  -- | Prefix for the Redis keys
+  keyPrefix :: BS.ByteString
+
+  -- | Which Redis instance the queue lives in.
+  type RedisInstance jq :: Instance
+  type RedisInstance jq = DefaultInstance
+
+-- | This queue contains many requests.
+-- There is only one request queue and it's read by all workers.
+data RequestQueue jq = RequestQueue
+instance JobQueue jq => Ref (RequestQueue jq) where
+  type RefInstance (RequestQueue jq) = RedisInstance jq
+  type ValueType (RequestQueue jq) = [(Priority, (UUID, Int, BSL.ByteString))]
```

Comment:
Would it be cleaner to replace these (nested) tuples with (nested?) records? Or maybe a record in a tuple?

Also below in `RunningJobs`. I imagine the triple part is actually shared and morally the same type (`Job`). 

Providing the names for the components of the triple would also be helpful when reading the code, as it's not clear to me at this point what the interpretation is. (That's just because I'm reading top-to-bottom but still, having extra clues would be nice.)

Reply:
Good idea, fixed